In [1]:
from decimal import Decimal
import decimal

### the decimal method doesnt stop only at confusing us on Its functionality, It also gives us other tools to represent a decimal number:

In [2]:
a = Decimal((1,(3,1,4,1,5),-4))
a

Decimal('-3.1415')

### we can use a tuple to represent a decimal to where the tuple is:
### (s,(d1,d2,d3,...),exp
### s : is the sign
### d's : are the digits in order
### exp: exponenet to the power of  10 of the number

## on our example a , 3.1415 is -31415 x 10e-2

#### string on decimal method still stands as a better option, most of the time

In [3]:
a = Decimal("-3.1414")
a

Decimal('-3.1414')

In [4]:
g_context = decimal.getcontext()
g_context

Context(prec=28, rounding=ROUND_HALF_EVEN, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[], traps=[InvalidOperation, DivisionByZero, Overflow])

In [5]:
g_context.prec

28

#### the prcision attribute in the getcontext method, is a range for how much significant digits will be taken into consideration when doing operations on decimal.Decimal() objects

In [22]:
a = Decimal("0.1235")
b = Decimal("0.1235")


In [28]:
g_context.prec = 2
a+b  # without the precision, a+b = 0.2470, but gets rounded to 0.25 since 0.24 is the first two significant digits 
     # and 7 is closer to the upper and than It is to the lower end

Decimal('0.25')

In [29]:
print(a)
print(b)

0.1235
0.1235


### precision attribute does not affect the decimal Itself, It only affect the operation, you can call It as an operation rounder

## we also know that global and local context are different
### wither we change the precision or another attribute, that only affects the inner decimals

In [1]:
import decimal
from decimal import Decimal
a = Decimal(0.123456789)
b = Decimal(0.123456789)
print(f" operation: {a+b}")
with decimal.localcontext() as l_ctx:
    l_ctx.prec = 2
    c = a + b
    print(f"value of c in local context {c}")
print(f"value of c in global context {c}")
print(f"operation {a+b}")

 operation: 0.2469135779999999946721089827
value of c in local context 0.25
value of c in global context 0.25
operation 0.2469135779999999946721089827


### and of course value of c is same as It is on local context, since we assigned the result into a variable, the variable doesnt just get rid of It

#### need to be careful with changing global context, almost got tricked again into a 3am session of why my code is doing this and not that